src/devices/device_models/thermal_generation.jl

```diff
@@ -245,19 +247,16 @@ function ramp_constraints(ps_m::CanonicalModel,
     if !isempty(data[2])
```

Comment:
This isn't new to this PR, but you have a bit of duplicated code in these ramp_constraints functions (it still is a recent change).  You could consolidate the logic in a function.  The only difference is the function device* that gets called. You could pass that in to the new function.